common/include/pcl/pcl_macros.h

```diff
@@ -448,3 +448,34 @@ aligned_free(void* ptr)
 #else
   #define PCL_IF_CONSTEXPR(x) if (x)
 #endif
+
+// [[unlikely]] can be used on any conditional branch, but __builtin_expect is restricted to the evaluation point
+// This makes it quite difficult to create a single macro for switch and while/if
+/**
+ * @def PCL_CONDITION_UNLIKELY
+ * @brief Tries to inform the compiler to optimize codegen assuming that the condition will probably evaluate to false
+ * @note Prefer using `PCL_{IF,WHILE}_UNLIKELY`
+ * @warning This can't be used with switch statements
+ * @details This tries to help the compiler optimize for the unlikely case.
+ * Most compilers assume that the condition would evaluate to true in if and while loops (reference needed)
+ * As such the opposite of this macro (PCL_CONDITION_LIKELY) will not result in significant performance improvement
+ * 
+ * Some sample usage:
+ * @code{.cpp}
+ * if PCL_CONDITION_UNLIKELY(x == 0) { return; } else { throw std::runtime_error("some error"); }
+ * //
+ * while PCL_CONDITION_UNLIKELY(wait_for_result) { sleep(1); }  // busy wait, with minimal chances of waiting
+ * @endcode
+ */
+#if __has_cpp_attribute(unlikely)
+  #define PCL_CONDITION_UNLIKELY(x) (static_cast<bool>(x)) [[unlikely]]
+#elif defined(__GNUC__)
+  #define PCL_CONDITION_UNLIKELY(x) (__builtin_expect(static_cast<bool>(x), 0))
+#elif defined(__clang__) && (PCL_LINEAR_VERSION(__clang_major__, __clang_minor__, 0) >= PCL_LINEAR_VERSION(3, 9, 0))
+  #define PCL_CONDITION_UNLIKELY(x) (__builtin_expect(static_cast<bool>(x), 0))
+#else  // MSVC has no such alternative
+  #define PCL_CONDITION_UNLIKELY(x) (x)
+#endif
+
+#define PCL_IF_UNLIKELY(x) if PCL_CONDITION_UNLIKELY(x)
```

Comment:
Our (google) benchmark functionality could be used, to compare a standard `if` against `PCL_IF_UNLIKELY` (the latter should lead to a speed improvement). [This line](https://github.com/PointCloudLibrary/pcl/blob/c51f7a2dfc2097e74957143e29a0310f3c1a5514/sample_consensus/include/pcl/sample_consensus/impl/ransac.hpp#L120) could be a nice candidate for a `PCL_IF_UNLIKELY`. But don't feel obligated to write such a benchmark, I just wanted to say it would be interesting to know how big of an impact this hint to the compiler can be :smile: